Task 13 - StepArrays

AlinaKlishyna opened the pull request.

AlinaKlishyna committed:
Created un method enterNumber()

AlinaKlishyna committed:
Created un methods createArray() and displayConsole()

Added data in main()

AlinaKlishyna committed:
Created method sortEvenOdd()

AlinaKlishyna committed:
Created method displayConsole(), sumElements() and minElementInRow()

AlinaKlishyna committed:
Modified methods: min(), minElementInRow()

AlinaKlishyna committed:
Added sortDecline(), sortGrowth() and swap()

AlinaKlishyna committed:
Created method absoluteMin()

AlinaKlishyna committed:
Modified method createArray()

AlinaKlishyna committed:
Modified methods absoluteMin(),minElementInRow(), createArray()

Added method countEmptyArrays()

AlinaKlishyna committed:
Deleted methods countEmptyArrays() and displayConsole()

Modified absoluteMin() and minElementInRow()

AlinaKlishyna committed:
Deleted methods countEmptyArrays() and displayConsole()

Modified absoluteMin() and minElementInRow()

AlinaKlishyna committed:
Merge branch 'Task13--Step-arrays'

# Conflicts:
#	src/com/gmail/alinakotova102/drafts/StepArrays.java

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+        int countRows = enterNumberBetween();
+        System.out.print("Enter M - the maximum number of elements in a row: ");
+        int maxCount = enterNumber();
+        while (!(maxCount >= 0)) {
```
Чому б тоді умову не замисати як `maxCount < 0`?

AlinaKlishyna replied:
```java
int maxCount = enterNumberBetween();
``` 
змінила таким чином, тепер цикл не потрібен(там є діапазон)

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+        System.out.print("Enter N - the number of lines: ");
+        int countRows = enterNumberBetween();
+        System.out.print("Enter M - the maximum number of elements in a row: ");
+        int maxCount = enterNumber();
```
Краще, щоб метод `enterNumber` приймав обмеження на ввод числа. Не розумію чому є метод для вводу чисел і ще якась додаткова логіка в `main`. Замість `while`, можливо, `do-while` доцільніше?
До того ж у тебе є окремий метод `enterNumberBetween`.

AlinaKlishyna replied:
```java
 int maxCount = enterNumberBetween();
``` 
змінила ✓

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+        int minRandom = enterNumber();
+        System.out.print("Max number: ");
+        int maxRandom = enterNumber();
+        while (maxRandom - minRandom == 1 || maxRandom == minRandom) {
```
А що буде у випадку коли `minRandom - maxRandom == 1`?

AlinaKlishyna replied:
```java
 while (Math.abs(maxRandom-minRandom) == 1 || maxRandom == minRandom ) {
            System.out.println("Error! There is no range between numbers\nMax number: ");
            maxRandom = enterNumber();
        }
```

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+        array = fillArray(array, minRandom, maxRandom);
+
+        System.out.println("\nInitial array");
+        displayConsole(array);
```
імхо, просто `display` було б достатньо. Або ж `print`

AlinaKlishyna replied:
```java
display()
``` 
змінила ✓

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+    }
+
+    public static boolean divisionCheck(int number, int divisionNumber) {
+        if (number % divisionNumber == 0) {
```
Тобто якщо умова істина, то ти повертаєш `true`, а якщо хибна `false`. Тоді твій метод можна записати так
```
public static boolean divisionCheck(int number, int divisionNumber) {
    return number % divisionNumber == 0;
}
```
Не потрібно нічого порівнювати з `true`/`false`. Та писати `if`, якщо потрібно повернути результат порівняння

пс: а цей метод точно потрібен?

AlinaKlishyna replied:
Це гарна порада! Й не помітила, що можна короче зробити..
Щодо надобності гарне питання - але трішки коду стало менше
```java
public static boolean divisionCheck(int number, int divisionNumber) {
        return number % divisionNumber == 0;
    }
```

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+            }
+        }
+        String message = "";
+        if (!(count > 0)) {
```
Дана умова еквівалентна `count <= 0`

AlinaKlishyna replied:
```java
if (count<=0) {
            display(array);
            return message = "Possible!";
        }
```

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+        int count = 0;
+        for (int i = 0; i < array.length; i++) {
+            for (int j = 0; j < array[i].length; j++) {
+                if (!(array[i].length == 0) && !(absoluteMin == 0) || array[i][j]!=0) {
```
Якщо довжина рядка 0, тобто `array[i].length == 0` - істина, то виконання просто не зайде у цикл по `j`. На першому кроці цього циклу `j == 0` і тоді умова `j < array[i].length` буде обчислена як `0 < 0` - хибно.
Далі... якщо `absoluteMin == 0`, то чому дану перевірку потрібно робити для кожного елемента. Хіба абсолютний мінімум може змінитись? Зроби перевірку один раз перед циклами.
`array[i][j]!=0` - а якщо елемент 0? мені здається 0 я можу ділити на будь-що, крім нуля (тобто якщо абс мін не 0, то нульвий елемент на нього я можу поділити)

AlinaKlishyna replied:
```java
 public static String divideByAbsoluteMin(int[][] array, int absoluteMin) {
        int count = 0;
        String message = "";
        if (!(absoluteMin == 0)) {
            for (int i = 0; i < array.length; i++) {
                if (!(array[i].length == 0)) {
                    for (int j = 0; j < array[i].length; j++) {
                        if (divisionCheck(array[i][j], absoluteMin)) {
                            array[i][j] /= absoluteMin;
                        } else {
                            count++;
                        }
                    }
                } else {
                    count++;
                }
            }
        } else {
            return message = "Impossible";
        }
        if (!(count > 0)) {
            display(array);
            return message = "Possible!";
        }
        return message = "Impossible";
    }
```

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+                int init = array[i][0];
+                for (int j = 0; j < array[i].length; j++) {
+                    if (array[i][j] < init) {
+                        init = min(array[i]);
```
`init` - це скорочення від initial? чому така назва? дану змінну логічно було б назвати `min`

AlinaKlishyna replied:
виправлено

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+                modArray[i] = NOT_DEFINE;
+            } else {
+                int init = array[i][0];
+                for (int j = 0; j < array[i].length; j++) {
```
Давай поясню що ти тут робиш: для і-го рядка, якщо його довжина не 0, ти будеш перебирати всі елементи (цикл по j). Якщо зустрінеться елемент, менше першого елемента в рядку, то ти викличеш метод `min`, який поверне мінімум в і-му рядку. Після цього ти продовжеш перебирати елементи у рядку. Навіщо? Чому не достатньо викликати просто метод `min`?

AlinaKlishyna replied:
public static Integer[] minElementInRow(int[][] array) {
        Integer[] modArray = new Integer[array.length];
        for (int i = 0; i < array.length; i++) {
            if (array[i].length == 0) {
                modArray[i] = NOT_DEFINE;
            } else {
                int min = array[i][0];
                for (int j = 0; j < array[i].length; j++) {
                    min = min(array[i]);
                }
                modArray[i] = min;
            }
        }
        return modArray;
    }

AlinaKlishyna replied:
```java
public static Integer[] minElementInRow(int[][] array) {
        Integer[] modArray = new Integer[array.length];
        for (int i = 0; i < array.length; i++) {
            if (array[i].length == 0) {
                modArray[i] = NOT_DEFINE;
            } else {
                int min = array[i][0];
                for (int j = 0; j < array[i].length; j++) {
                    min = min(array[i]);
                }
                modArray[i] = min;
            }
        }
        return modArray;
    }
```

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+        int sum = 0;
+        for (int i = 0; i < array.length; i++) {
+            for (int j = 0; j < array[i].length; j++) {
+                sum += Math.abs(array[i][j]);
```
За умовою потрібно суму елементів знайти, а не суму модулів. Навіщо тут `Math.abs()`?

AlinaKlishyna replied:
```java
public static int sumElements(int[][] array) {
        int sum = 0;
        for (int i = 0; i < array.length; i++) {
            for (int j = 0; j < array[i].length; j++) {
                sum += array[i][j];
            }
        }
        return sum;
    }
```

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+        return sum;
+    }
+
+    public static int[][] swap(int[][] array, int row, int column, int newColumn) {
```
Навіщо тут стільки параметрів? Якщо потрібно обміняти елементи у рядку, то метод мавби мати наступну сігнатуру
```
int[][] swap(int[] row, int i, int j) { ... }
```

AlinaKlishyna replied:
```java
 public static int[] swap(int[] array, int num, int num2 ) {
            int saveNumber = array[num + 1];
            array[num + 1] = array[num2];
            array[num2] = saveNumber;
        return array;
    }

    public static int[][] sortGrowth(int[][] array, int index) {
        for (int j = 0; j < array[index].length - 1; j++) {
            for (int k = 0; k < array[index].length - 1; k++) {
                if (array[index][k] > array[index][j + 1]) {
                    swap(array[index], j, k);
                }
            }
        }
        return array;
    }

    public static int[][] sortDecline(int[][] array, int index) {
        for (int j = 0; j < array[index].length - 1; j++) {
            for (int k = 0; k < array[index].length - 1; k++) {
                if (array[index][k] < array[index][j + 1]) {
                    swap(array[index], j,k);
                }
            }
        }
        return array;
    }
```

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+        return array;
+    }
+
+    public static int[][] sortGrowth(int[][] array, int index) {
```
Ти сортуєш одновимірний масив тут. Тоді зроби метод з такою сигнатурою
```
public static int[][] sortAscending(int[] row) { ... }
```
Метод буде приймати рядок і сортувати його. Ascending/Descending або Asc/Desc (це загально прийняті скорочення), імхо, краще ніж Growth/Decline.

AlinaKlishyna replied:
```java
public static int[] sortGrowth(int[] array) {
        for (int index = 0; index < array.length; index++) {
            for (int j = 0; j < array.length - 1; j++) {
                for (int k = 0; k < array.length - 1; k++) {
                    if (array[k] > array[j + 1]) {
                        swap(array, j, k);
                    }
                }
            }
        }
        return array;
    }

    public static int[] sortDecline(int[] array) {
        for (int index = 0; index < array.length; index++) {
            for (int j = 0; j < array.length - 1; j++) {
                for (int k = 0; k < array.length - 1; k++) {
                    if (array[k] < array[j + 1]) {
                        swap(array, j,k);
                    }
                }
            }
        }
        return array;
    }

    public static int[][] sortEvenOdd(int[][] array) {
        for (int i = 0; i < array.length; i++) {
            if ((i + 1) % 2 == 0) {
                sortGrowth(array[i]);
            } else {
                sortDecline(array[i]);
            }
        }
        return array;
    }
```

AlinaKlishyna replied:
```java
public static int[] sortAsc(int[] array) {
        for (int index = 0; index < array.length; index++) {
            for (int j = 0; j < array.length - 1; j++) {
                for (int k = 0; k < array.length - 1; k++) {
                    if (array[k] > array[j + 1]) {
                        swap(array, j, k);
                    }
                }
            }
        }
        return array;
    }

    public static int[] sortDesc(int[] array) {
        for (int index = 0; index < array.length; index++) {
            for (int j = 0; j < array.length - 1; j++) {
                for (int k = 0; k < array.length - 1; k++) {
                    if (array[k] < array[j + 1]) {
                        swap(array, j,k);
                    }
                }
            }
        }
        return array;
    }

    public static int[][] sortEvenOdd(int[][] array) {
        for (int i = 0; i < array.length; i++) {
            if ((i + 1) % 2 == 0) {
                sortAsc(array[i]);

            } else {
                sortDesc(array[i]);
            }
        }
        return array;
    }
```

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+        return array;
+    }
+
+    public static int[][] sortDecline(int[][] array, int index) {
```
Ті ж зауваження, що і для методу вище

AlinaKlishyna replied:
виправила

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+    public static void displayConsole(Integer[] array) {
+        System.out.print("[");
+        int count = 0;
+        for (Integer element : array) {
```
Якщо тобі потрібен індекс елемента (щоб розуміти останній це елемент у масиві чи ні), то чому не використати звичайний `for`? Які переваги у `for each` в даному випадку
У тебе в різних частинах програми array має різний сенс. Десь це матриця, тут це одновимірний масив. Вживання одного терміну у різних значеннях заплутує

AlinaKlishyna replied:
```java
public static void display(int[][] matrix) {
        for (int i = 0; i < matrix.length; i++) {
            System.out.print(Arrays.toString(matrix[i]) + "\n");
        }
        System.out.println();
    }

    public static void display(Integer[] array) {
        System.out.print("[");
        for (int i = 0; i < array.length; i++) {
            System.out.print(array[i]);
            if (i<array.length-1){
                System.out.print(", ");
            }
        }
        System.out.println("]");
    }
``` 
виправила назви на matrix, де є "int[][]"

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+    public static int[][] createArray(int countInRow, int length) {
+        int[][] numbersRandom = new int[countInRow][];
+        for (int i = 0; i < numbersRandom.length; i++) {
+            if (length == 0) {
```
Мабуть, тут мала б бути перевірка `length <= 0`?

AlinaKlishyna replied:
точно!

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+        return numbersRandom;
+    }
+
+    public static int[][] fillArray(int[][] numbersRandom, int minRandom, int maxRandom) {
```
Чому ж `numbersRandom` коли ти матрицю заповнюєш. Назви цю змінну `matrix` тощо. В даному випадку змінні `numbersRandom`, `minRandom`, `maxRandom` сприймаються як `numbers`, `min`, `max`

AlinaKlishyna replied:
виправила

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+
+    public static int absoluteMin(int[][] array) {
+        Integer[] arrayMinNumbers = minElementInRow(array);
+        int init = arrayMinNumbers[0];
```
`init` - невдала назва для змінною. По факту це мінімум.

AlinaKlishyna replied:
![image](https://github.com/AlinaKlishyna/KlishynaAlina_JavaBase_03_02_2024/assets/48547022/2212e098-1d6d-4d5d-b120-812cb937b8dd)

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+        return message = "Impossible";
+    }
+
+    public static int absoluteMin(int[][] array) {
```
Ти ж вже знайшла мінімальні елементи в рядках. Чому б не перевикористати логіку і не шукати абсолютний мінімум серед знайдених мінімумів?

AlinaKlishyna replied:
```java
 public static boolean divisionCheck(int number, int divisionNumber) {
        return number % divisionNumber == 0;
    }
```

JIeIIIa replied:
Щось твій коментар не того зауваження стосується

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,247 @@
+        System.out.println("]");
+    }
+
+    public static int[][] createArray(int countInRow, int length) {
```
`countInRow` - це ж не кількість в рядку. Це мало б бути `rowCount`

AlinaKlishyna replied:
виправила

JIeIIIa reviewed (changes requested):
Як поведе себе твоя програма, якщо користувач вкаже 5 рядків і максимум у кожному рядку 0?

AlinaKlishyna committed:
Created class Person and methods main() and personInfo()

AlinaKlishyna committed:
Reformatted Code

AlinaKlishyna committed:
Modified variable lastname -> lastName

AlinaKlishyna committed:
Added class PersonRunner

Method main() in class was modified

In class Person added getters and setters. Modified method personInfo()

AlinaKlishyna committed:
Reformatted Code

AlinaKlishyna committed:
Reformatted Code

AlinaKlishyna committed:
Reformatted Code

AlinaKlishyna committed:
Modified methods

AlinaKlishyna committed:
Delete src/com/gmail/alinakotova102/task14/Person.java

AlinaKlishyna committed:
Delete src/com/gmail/alinakotova102/task14/PersonRunner.java

AlinaKlishyna committed:
Delete src/com/gmail/alinakotova102/drafts/StepArrays.java

AlinaKlishyna committed:
Revert "Reformatted variables num -> numberFirst, num2 -> numberSecond"

This reverts commit 08d09b2d9530542db11460e0c5970b2dd9c5d694.

# Conflicts:
#	src/com/gmail/alinakotova102/drafts/StepArrays.java

AlinaKlishyna committed:
Deleted method personInfo() with arguments

Modified method main() in class PersonRunner

AlinaKlishyna committed:
Merge remote-tracking branch 'origin/Task14-personInfo()' into Task14-personInfo()

# Conflicts:
#	src/com/gmail/alinakotova102/drafts/StepArrays.java
#	src/com/gmail/alinakotova102/task14/Person.java
#	src/com/gmail/alinakotova102/task14/PersonRunner.java

AlinaKlishyna committed:
Deleted method personInfo() with arguments

Modified method main() in class PersonRunner

AlinaKlishyna committed:
test-commit

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,251 @@
+    public static void main(String[] args) {
+        System.out.println("Creation of a stepped array with random numbers");
+        System.out.print("Enter N - the number of lines: ");
+        int countRows = enterNumberBetween();
```
Тобі ж насправді не потрібно два методи для вводу чисел. Зроби один із наступною сигнатурою:
```
public static int enterNumberBetween(int ordinary, int bound) {
}
```
В даний метод ти будеш передавати межі, в яких потрібно ввести число. Для кіл-сті рядків/стовпчиків це буде від 0 до Integer.MAX_INT (чи яка верхня межа має бути). А для значень елементів 1..999 (так, як у тебе зараз)

AlinaKlishyna replied:
```java
 public static int enterNumberBetween(int ordinary, int bound) {
        Scanner scanner = new Scanner(System.in);
        while (true) {
            if (scanner.hasNextInt()) {
                int number = scanner.nextInt();
                if (number >= ordinary && number < bound) {
                    return number;
                } else {
                    System.out.print("Enter a number between [1-999]: ");
                }
            } else {
                System.out.print("Value non valid! Enter again: ");
            }
            scanner.nextLine();
        }
    }
``` 

```java
  public static void main(String[] args) {
        System.out.println("Creation of a stepped array with random numbers");
        System.out.print("Enter N - the number of lines: ");
        int countRows = enterNumberBetween(0, Integer.MAX_VALUE);
        System.out.print("Enter M - the maximum number of elements in a row: ");
        int maxCount = enterNumberBetween(1, 999);

        System.out.print("\nEnter number for create array with random numbers\nMin number: ");
        int minRandom = enterNumberBetween(-999, 999);
        System.out.print("Max number: ");
        int maxRandom = enterNumberBetween(-999, 999);
        while (Math.abs(maxRandom - minRandom) == 1 || maxRandom == minRandom) {
            System.out.println("Error! There is no range between numbers\nMax number: ");
            maxRandom = enterNumberBetween(-999, 999);
        }
``` 
Я змінила і для значення рандомів

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,251 @@
+        display(minElement);
+
+        int absoluteMinElement = absoluteMin(array);
+        System.out.println("\nAbsolute minimal element: " + absoluteMinElement);
```
Вивід твоєї програми коли кіл-сть рядків 5, а макс елементів в рядку менше 1
```
Minimal element in row: [null, null, null, null, null]

Absolute minimal element: 0
```
`null` - ти виводиш коли мінімума не існує. Для користувача краще б було замість цього вивести якийсь інший текст, але хай вже так лишається.
Проблема в іншому: чому для кожного окремого рядка мінімум не існує, а абсолютний мінімум при цьому є і він дорівнює 0? Я очікую, що в даному випадку абсолютний мінімум також не мав би існувати.

AlinaKlishyna replied:
```java
 public static Integer absoluteMin(int[][] matrix) {
        if (min(minElementInRow(matrix)) == 0) {
            return NOT_DEFINE;
        }
        return min(minElementInRow(matrix));
    }
``` 
![image](https://github.com/AlinaKlishyna/KlishynaAlina_JavaBase_03_02_2024/assets/48547022/1683b0ef-e1df-4ff0-b6c3-6903e0958984)

JIeIIIa replied:
Я запустив твій код з гілки main і отримав NPE:
```
Creation of a stepped array with random numbers
Enter N - the number of lines: 4
Enter M - the maximum number of elements in a row: 1

Enter number for create array with random numbers
Min number: 0
Max number: 2

Initial array
[]
[]
[]
[]

Modified array
[]
[]
[]
[]

Sum of elements: 0

Minimal element in row: [null, null, null, null]
Exception in thread "main" java.lang.NullPointerException: Cannot invoke "java.lang.Integer.intValue()" because "minValue" is null
	at com.gmail.alinakotova102.drafts.StepArrays.min(StepArrays.java:82)
	at com.gmail.alinakotova102.drafts.StepArrays.absoluteMin(StepArrays.java:65)
	at com.gmail.alinakotova102.drafts.StepArrays.main(StepArrays.java:43)
```

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,251 @@
+    }
+
+    public static String divideByAbsoluteMin(int[][] matrix, int absoluteMin) {
+        int count = 0;
```
Чому для того, щоб виконати ділення ти ще рахуєш якусь кількість?
На початку метода ти можен зробити перевірки: якщо абсолютний мінімум не існує або ж він 0, то  надрукувати у консоль помилку і зробити return
```
if ( .... ) {
    System.out.println("Error");
    return;
}
for (... ) {
   ...
}

AlinaKlishyna replied:
```java
public static String divideByAbsoluteMin(int[][] matrix, Integer absoluteMin) {
        String message = "";
        if (absoluteMin != null) {
            for (int i = 0; i < matrix.length; i++) {
                if (!(matrix[i].length == 0)) {
                    for (int j = 0; j < matrix[i].length; j++) {
                        matrix[i][j] /= absoluteMin;
                    }
                } else {
                    return message = "Impossible";
                }
            }
        } else {
            return message = "Impossible";
        }
        display(matrix);
        return message = "Possible";
    }
```

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,251 @@
+    public static String divideByAbsoluteMin(int[][] matrix, int absoluteMin) {
+        int count = 0;
+        String message = "";
+        if (!(absoluteMin == 0)) {
```
Так робити не потрібно, але для розуміння: твої умови можна і потрібно було писати простіше.
```
if (absoluteMin != 0) {
    for (int i = 0; i < matrix.length; i++) {        
        for (int j = 0; j < matrix[i].length; j++) {
            if (divisionCheck(matrix[i][j], absoluteMin)) {
                matrix[i][j] /= absoluteMin;
            } 
        }
    }
    display(matrix);
    return"Possible!";
} else {
    return "Impossible";
}
```

AlinaKlishyna replied:
```java
 public static String divideByAbsoluteMin(int[][] matrix, Integer absoluteMin) {
        if (absoluteMin != null) {
            for (int i = 0; i < matrix.length; i++) {
                for (int j = 0; j < matrix[i].length; j++) {
                    matrix[i][j] /= absoluteMin;
                }
            }
            display(matrix);
            return "Possible";
        } else {
            return "Impossible";
        }
    }
```

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,251 @@
+        System.out.print(divideByAbsoluteMin(array, absoluteMinElement));
+    }
+
+    public static boolean divisionCheck(int number, int divisionNumber) {
```
Я щось в перший раз пропустив, але... а чому ти взагалі перевіряєш, що одне число можна націло поділити на інше? Ми виконуємо операцію цілочисельного ділення в будь-якому випадку якщо можемо це виконати. Тобто коли абсолютний мінімум існує і він не 0.

AlinaKlishyna replied:
```java
public static String divideByAbsoluteMin(int[][] matrix, Integer absoluteMin) {
        int count = 0;
        String message = "";
        if (absoluteMin != null) {
            for (int i = 0; i < matrix.length; i++) {
                if (!(matrix[i].length == 0)) {
                    for (int j = 0; j < matrix[i].length; j++) {
                        matrix[i][j] /= absoluteMin;
                    }
                } else {
                    count++;
                }
            }
        } else {
            return message = "Impossible";
        }
        if (count <= 0) {
            display(matrix);
            return message = "Possible!";
        }
        return message = "Impossible";
    }
```

AlinaKlishyna replied:
```java
public static String divideByAbsoluteMin(int[][] matrix, Integer absoluteMin) {
        int count = 0;
        String message = "";
        if (absoluteMin != null) {
            for (int i = 0; i < matrix.length; i++) {
                if (!(matrix[i].length == 0)) {
                    for (int j = 0; j < matrix[i].length; j++) {
                        matrix[i][j] /= absoluteMin;
                    }
                } else {
                    count++;
                }
            }
        } else {
            return message = "Impossible";
        }
        if (count <= 0) {
            display(matrix);
            return message = "Possible!";
        }
        return message = "Impossible";
    }
``` 
Удалила метод divisionCheck()

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,251 @@
+            minValue = 0;
+        }
+        for (int i = 0; i < array.length; i++) {
+            if (array[i] == null) {
```
Тобто якщо елемент масива `null`, то йому потрібно присвоїти нове значення, яке дорівнює `null`. Даний блок не має сенсу

AlinaKlishyna replied:
```java
  public static int min(Integer[] array) {
        int minValue = array[0];
        for (int i = 0; i < array.length; i++) {
            if (array[i] == null) {
                array[i] = NOT_DEFINE;
            } else {
                if (array[i] < minValue) {
                    minValue = array[i];
                }
            }
        }
        return minValue;
    }
```

JIeIIIa replied:
Так зауваження не виправлено( Ще раз: у тебе є масив `array = {null, 1, 2, 3}`. Починає працювати цикл `for` по `i`. На першому кроці циклу буде:
```
Integer minValue = array[0];
for (int i = 0; i < array.length; i++) {
    if (array[i] == null) {  // i == 0 and array[0] - це null. умова істина
        array[i] = NOT_DEFINE; // тут ти для array[0] присвоюєш значення null. але ж воно і так мал це значення. навіщо тоді це робити???
    }
```

AlinaKlishyna committed:
Merge branch 'main' of https://github.com/AlinaKlishyna/KlishynaAlina_JavaBase_03_02_2024 into Task14-personInfo()

# Conflicts:
#	src/com/gmail/alinakotova102/task14/Person.java
#	src/com/gmail/alinakotova102/task14/PersonRunner.java

AlinaKlishyna committed:
Merge pull request #16 from AlinaKlishyna/Task14-personInfo()

Task14 person info()

AlinaKlishyna committed:
Deleted methods enterNumber()

Modified method enterNumberBetween()

Reformatted code

AlinaKlishyna committed:
Modified int -> Integer

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,215 @@
+            if (matrix[i].length == 0) {
+                modArray[i] = NOT_DEFINE;
+            } else {
+                int min = matrix[i][0];
```
В блоці `else`  достатньо було б просто викликати метод пошуку мінімума. Так:
```
else {
    modArray[i] = min(matrix[i]);
}
```
Натомість що у тебе ваівдбувається: нехай є матриця matrix = { {4, 3, 2, 1}, {5, 6}}. Розглянемо блок `else` для першого рядка матриці:
```
    int min = matrix[i][0]; // тут буде 4 (бо i == 0)
    for (int j = 0; j < matrix[i].length; j++) {
        min = min(matrix[i]); // на першому кроці циклу, коли j == 0 ти викликаєш мінімум для {4, 3, 2, 1}. буде повернуто 1. 
       // мінімум ти вже знайшла, але ж цикл for ще не завершено. тому для j == 1, j == 2, j == 3 ти так само викликаєш метод min({4, 3, 2, 1}) - це зайва робота яка негативно впливає на час виконання. замість складності O(n) у тебе виходить складність O(n^2)
    }
    modArray[i] = min;
```

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,215 @@
+    }
+
+    public static Integer[] sortAsc(Integer[] array) {
+        for (int index = 0; index < array.length; index++) {
```
Цикли по `j` та `k` - це реалізація бульбашкового сортування. а навіщо ще цикл по `index`? змінна `index` взагалі ніде не фігурує в методі. Якщо довжина масива array довірнює `N`, то ти сортуєш цей масив `N` разів? Складність методу `sortAsc` у тебе вийшла O(n^3). Мала б бути O(n^2)

JIeIIIa commented on src/com/gmail/alinakotova102/drafts/StepArrays.java:
```diff
@@ -1,7 +1,215 @@
+        for (int index = 0; index < array.length; index++) {
+            for (int j = 0; j < array.length - 1; j++) {
+                for (int k = 0; k < array.length - 1; k++) {
+                    if (array[k] > array[j + 1]) {
```
В Bubble Sort порівнюються сусідні елементи. Мало б бути
```
if (array[k] > array[k + 1]) {
}
```

AlinaKlishyna committed:
Created classes Burger and BurgerMain

AlinaKlishyna committed:
Merge branch 'Task15-Burgers'

AlinaKlishyna committed:
Modified constructors

AlinaKlishyna committed:
Deleted method info() and modified constructor Burger()

AlinaKlishyna committed:
Reformatted code

AlinaKlishyna committed:
Added new constructors

AlinaKlishyna committed:
Reformatted code

AlinaKlishyna committed:
Modified method info() in class Burger

AlinaKlishyna committed:
Merge pull request #18 from AlinaKlishyna/Task15-Burgers

Task15 burgers
